Small improvements in Raft

mmaslankaprv opened the pull request:
Added metrics allowing estimating the size of the append entries buffer. This may be helpful to validate how well Raft coalesce writes and flushes. The average size of the buffer can be calculated as the ratio between the number of append entries requests and append entries buffer flushes executed in the period of time.

Additional small optimizations in Raft implementation.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [x] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
r/append_entries_buffer: added metrics to the buffer

Added metrics allowing estimating the size of the append entries buffer.
This may be helpful to validate how well Raft coalesce writes and
flushes. The average size of the buffer can be calculated as the ratio
between the number of append entries requests and append entries buffer
flushes executed in the period of time.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/service: use async_for_each to dispatch heartbeats

Minimize number of times `maybe_yield` is called by using async for each
utility function.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--62af78bc-fc34-43dc-b6ab-f8236e52ecec-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60950">build#60950</a></summary>


| test_id                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60950#01948376-cdfe-47e0-b500-709f9516046c | FLAKY         | 1/2      |
</details>
<!--/body-->

mmaslankaprv committed:
r/consensus: return const ref to configuration from `config()`

Signed-off-by: Michał Maślanka <michal@redpanda.com>

dotnwat commented on src/v/raft/service.h:
```diff
@@ -137,16 +138,23 @@ class service final : public raftgen_service {
+              shard_replies.full_heartbeats.begin(),
+              shard_replies.full_heartbeats.end(),
+              [&reply](full_heartbeat_reply& full_reply) {
+                  reply.add(full_reply.group, full_reply.result);
```
why does the reply not contain `full_reply.data`?

mmaslankaprv replied:
great catch

mmaslankaprv committed:
r/tests: allow more memory for raft reconfiguration test

The `reconfiguration` test consumes more memory than it is allowed to be
allocated right now. Made the memory allowance larger to make sure the
test is not flaky.

Signed-off-by: Michał Maślanka <michal@redpanda.com>
